primitives/src/merkle_tree/merkle_tree_traits.rs

```diff
@@ -0,0 +1,261 @@
+    fn as_slice(&self) -> Vec<F> {
+        vec![F::from(*self)]
+    }
+
+    fn slice_len() -> usize {
+        1
+    }
+}
+
+/// A merkle commitment consists a root hash value, a tree height and number of
+/// leaves
+#[derive(
+    Eq, PartialEq, Clone, Copy, CanonicalSerialize, CanonicalDeserialize, Serialize, Deserialize,
+)]
+pub struct MerkleCommitment<I: IndexType<F>, F: Field> {
+    /// Root of a tree
+    pub root_value: F,
+    /// Height of a tree
+    pub height: usize,
+    /// Number of leaves in the tree
+    pub num_leaves: I,
+}
+
+/// Basic functionalities for a merkle tree implementation. Abstracted as an
+/// accumulator for fixed-length array. Supports generate membership proof at a
+/// given position and verify a membership proof.
+pub trait MerkleTree<F: Field>: Sized {
+    /// Merkle tree element type
+    type ElementType: ElementType<F>;
+    /// Hash algorithm used in merkle tree
+    type Digest: DigestAlgorithm<F>;
+    /// Index type for this merkle tree
+    type IndexType: IndexType<F>;
+    /// Leaf arity
+    type LeafArity: Unsigned;
+    /// Non-leaf arity
+    type TreeArity: Unsigned;
+    /// Merkle proof
+    type MembershipProof;
+    /// Batch proof
+    type BatchMembershipProof;
+
+    /// Construct a new merkle tree with given height
+    fn new(height: usize) -> Self;
+
+    /// Construct a new merkle tree with given height from a data slice
+    fn from_data(
+        height: usize,
+        data: impl IntoIterator<Item = impl Borrow<Self::ElementType>>,
+    ) -> Result<Self, PrimitivesError>;
+
+    /// Return the height of this merkle tree
+    fn height(&self) -> usize;
+    /// Return the maximum allowed number leaves
+    fn capacity(&self) -> Self::IndexType;
+    /// Return the current number of leaves
+    fn num_leaves(&self) -> Self::IndexType;
+
+    /// Return the current root value
+    fn root(&self) -> F;
+
+    /// Return a merkle commitment
+    fn commitment(&self) -> MerkleCommitment<Self::IndexType, F>;
+
+    /// Returns the leaf value given a position
+    /// * `pos` - zero-based index of the leaf in the tree
+    /// * `returns` - Leaf value at the position along with a proof.
+    ///   LookupResult::EmptyLeaf if the leaf position is empty or invalid,
+    ///   LookupResult::NotInMemory if the leaf position has been forgotten.
+    fn lookup(
+        &self,
+        pos: Self::IndexType,
+    ) -> LookupResult<Self::ElementType, Self::MembershipProof>;
+
+    /// Verify an element is a leaf of a Merkle tree given the proof
+    /// * `pos` - zero-based index of the leaf in the tree
+    /// * `proof` - a merkle tree proof
+    /// * `returns` - Ok(true) if the proof is accepted, Ok(false) if not. Err()
+    ///   if the proof is not well structured, E.g. not for this merkle tree.
+    fn verify(
+        &self,
+        pos: Self::IndexType,
+        proof: impl Borrow<Self::MembershipProof>,
+    ) -> Result<bool, PrimitivesError>;
```

Comment:
Should we have `pos` as input param? Intuitively just a proof is enough and practically, the pos info is embedded in `proof`.